wxUniv: Fix universal build under windows

stahta01 opened the pull request:
Add standard defines of wxUSE_WINRT and wxUSE_DBGHELP to wx/univ/setup.h
And, set wxUSE_TASKBARBUTTON to zero in wx/univ/setup.h
Also, disable wxUSE_GRAPHICS_CONTEXT in wx/msw/chkconf.h

This problem was discussed on [forums.wxwidgets.org](https://forums.wxwidgets.org/viewtopic.php?f=19&t=48256&p=205889)

Tim S.

stahta01 committed:
wxUniv: Fix universal build under windows

Add standard define of wxUSE_DBGHELP to wx/univ/setup.h
And, set wxUSE_WINRT and wxUSE_TASKBARBUTTON to zero in wx/univ/setup.h
Also, enable wxUSE_GRAPHICS_GDIPLUS in wx/msw/chkconf.h

vadz commented on include/wx/msw/chkconf.h:
```diff
@@ -225,6 +225,14 @@
 #   endif /* wxUSE_ACTIVEX */
 #endif /* __WINE__ */
 
+#if wxUSE_GRAPHICS_CONTEXT && defined(__WXUNIVERSAL__)
```
I don't really mind merging this to fix the build, but I'd like to mention that there is no reason wxGC shouldn't be available in wxUniv. I.e. the real fix would be fix the linking error instead.

stahta01 replied:
I will see if I can find the cause; do you know the name of the likely library names that are either missing or in the wrong place?

Tim S.

vadz replied:
No, sorry, but the libraries should be easy to determine from the linker error -- what is it exactly?

stahta01 replied:
I will have to remove the fix and rebuild; be at least a few hours to rebuild and find the error.
Tim S.

stahta01 replied:
Found the file I saved the error it was src/common/graphcmn.cpp:984: undefined reference to `wxGraphicsRenderer::GetDefaultRenderer()'

I will see if I can find the library name from that.

Tim S.

stahta01 replied:
Thinking about it; I am afraid the problem is not likely a missing library but missing code.

vadz replied:
This is not due to a missing library, but seems to be due to `wxUSE_GRAPHICS_GDIPLUS` being 0, so that this function is not getting defined in `src/msw/graphics.cpp` (the only alternative I see is that this file is not being compiled, but it should be).

So I guess we need to define this one, and probably `wxUSE_GRAPHICS_DIRECT2D` too (because why not) in `wx/univ/setup.h`. But actually I have no idea why we don't include all MSW-specific options in it automatically... probably just because nobody could be bothered to do it.

stahta01 replied:
I just edited and forced push my last commit.

Tim S.

vadz replied:
P.S. If you'd like to make this work reliably, you need to modify `build/update-setup-h` to include the contents of `include/wx/msw/setup_inc.h` into `include/wx/univ/setup.h`, like it's already done for several other files. Currently this is not done and instead we have a quite useless `include/wx/univ/setup_inc.h` which could be just removed and folded directly into `include/wx/univ/setup.h` itself because it's not needed everywhere else. Then any new wxMSW-specific options would be automatically added for wxUniv when `build/update-setup-h` is rerun.

stahta01 replied:
I will see about that as a possible long term task; no time right now.
Tim S.

stahta01 replied:
Please read my updated comment do I need to add to it?
I missed your post about wxUSE_GRAPHICS_GDIPLUS needing to be defined.
So, do I need to update the comment about that?

Tim S.

stahta01 replied:
I figure I should rebuild to verify none of the minor changes breaks the windows build.
Tim S.

stahta01 replied:
I just realized that I need to try it since, it appears to be a easy fix if just adding wxUSE_GRAPHICS_GDIPLUS fixes the problem.

stahta01 replied:
I just did a forced push I hope I am correct that under Windows "wxUSE_GRAPHICS_CONTEXT requires wxUSE_GRAPHICS_GDIPLUS". I added an error message that states that.

Tim S.

vadz commented on include/wx/univ/setup.h:
```diff
@@ -1637,6 +1637,19 @@
+//
+// Recommended setting: 1
+#if defined(_MSC_VER) && _MSC_VER >= 1700 && !defined(_USING_V110_SDK71_)
+    #define wxUSE_WINRT 1
```
OTOH I don't think it makes sense to enable this for wxUniv, this is used for native notifications that probably shouldn't be used by wxUniv.

stahta01 replied:
So, would adding it with a value of 0 to setup.h be correct?
Or editing the wx/msw/chkconf.h to set it to zero?

Tim S.

vadz replied:
Either would be fine.

stahta01 replied:
Okay, I have updated the pull request to set it to zero in setup.h and will be working on finding the linking error with core library. Laptop is getting hot so may have to stop for a few hours.

Tim S.

vadz reviewed (commented):
Thanks for looking at this, we can merge it in the current state if this fixes the build, as it would be already an improvement, but I don't think this is quite correct. Notably, wxGC really should be available in wxUniv, if we ever write any new code for it, it should use wxGC and not wxDC...

vadz commented on include/wx/msw/chkconf.h:
```diff
@@ -472,6 +472,20 @@
+#if wxUSE_GRAPHICS_CONTEXT && defined(__WXUNIVERSAL__) && !defined(wxUSE_GRAPHICS_GDIPLUS)
+#   define wxUSE_GRAPHICS_GDIPLUS 1
+#endif
+#if !wxUSE_GRAPHICS_GDIPLUS
+#   if wxUSE_GRAPHICS_CONTEXT
+#       ifdef wxABORT_ON_CONFIG_ERROR
+#           error "wxUSE_GRAPHICS_CONTEXT requires wxUSE_GRAPHICS_GDIPLUS"
+#       else
+#           undef wxUSE_GRAPHICS_GDIPLUS
+#           define wxUSE_GRAPHICS_GDIPLUS 1
+#       endif
+#   endif
+#endif
```
This is not quite right, you could have `wxUSE_GRAPHICS_DIRECT2D=1` and `wxUSE_GRAPHICS_GDIPLUS=0`, I think.

stahta01 replied:
Okay, I will do more changes. Tim S.

vadz commented on include/wx/msw/chkconf.h:
```diff
@@ -472,6 +472,20 @@
 #   define wxUSE_POSTSCRIPT 1
 #endif
 
+#if wxUSE_GRAPHICS_CONTEXT && defined(__WXUNIVERSAL__) && !defined(wxUSE_GRAPHICS_GDIPLUS)
+#   define wxUSE_GRAPHICS_GDIPLUS 1
+#endif
```
This should be done in `wx/univ/setup.h`.

vadz reviewed (changes requested):
Sorry, I don't understand what has happened here, but this is actually worse than the initial version.

If you want the simplest possible fix, please define `wxUSE_GRAPHICS_GDIPLUS` and `wxUSE_GRAPHICS_DIRECT2D` in `wx/univ/setup.h` (in addition to the other changes there).

You could add a consistency check to `wx/msw/chkconf.h` if you want, but this needs to account for D2D case too, if you do it at all.

stahta01 committed:
wxUniv: Change universal build under windows

Remove recent changes to wx/msw/chkconf.h
And, add sub set of "Windows-specific backends choices" to "wx/univ/setup.h"

vadz commented:
Thanks, this looks fine to me, so if this works for you, let's just merge it.
